services/api-server/src/simcore_service_api_server/api/routes/solvers.py

```diff
@@ -0,0 +1,179 @@
+logger = logging.getLogger(__name__)
+
+router = APIRouter()
+
+
+## SOLVERS -----------------------------------------------------------------------------------------
+
+
+@router.get("", response_model=List[Solver])
+async def list_solvers(
+    _catalog_client: CatalogApi = Depends(get_api_client(CatalogApi)),
+    url_for: Callable = Depends(get_reverse_url_mapper),
+):
+    """
+    Returns a list of the latest version of each solver
+    """
+    latest_solvers = [
+        Solver(
+            solver_url=url_for(
+                "get_solver_by_id",
+                solver_id=data["uuid"],
+            ),
+            **data,
+        )
+        for data in the_fake_impl.solvers
+    ]
+
+    return sorted(latest_solvers, key=attrgetter("name"))
+
+
+@router.get("/{solver_id}", response_model=Solver)
+async def get_solver_by_id(
+    solver_id: UUID,
+    url_for: Callable = Depends(get_reverse_url_mapper),
+):
+    try:
+        data = the_fake_impl.get(str(solver_id))
+        return Solver(
+            solver_url=url_for(
+                "get_solver_by_id",
+                solver_id=data["uuid"],
+            ),
+            **data,
+        )
+    except KeyError as err:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND) from err
+
+    # catalog get /image_id
+    # raise NotImplementedError(f"GET solver {solver_id}")
+
+
+@router.get("/{solver_name:path}/{version}", response_model=Solver)
+async def get_solver_by_name_and_version(
+    solver_name: SolverImageName,
+    version: str,
+    url_for: Callable = Depends(get_reverse_url_mapper),
+):
+    try:
+        if version == LATEST_VERSION:
+            data = the_fake_impl.get_latest(solver_name)
+        else:
+            data = the_fake_impl.get2(solver_name, version)
+
+    except KeyError as err:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND) from err
+
+    return Solver(
+        solver_url=url_for(
+            "get_solver_by_id",
+            solver_id=data["uuid"],
+        ),
+        **data,
+    )
+
+    # catalog get / key:latest
+    # raise NotImplementedError(f"GET {solver_name}:{version}")
+    # raise NotImplementedError(f"GET latest {solver_name}")
+
+
+async def _list_solvers_impl(
+    catalog_client: CatalogApi = Depends(get_api_client(CatalogApi)),
+    url_for: Callable = Depends(get_reverse_url_mapper),
+):
+    # TODO: pagination
+    # TODO: deduce user
+    # filter and format as in docker listings ??
+
+    # FIXME: temporary hard-coded user!
+    user_id = 1
+
+    available_services = await catalog_client.get(
+        "/services",
+        params={"user_id": user_id, "details": False},
+        headers={"x-simcore-products-name": "osparc"},
+    )
+
+    # TODO: move this sorting down to database?
+    # Create list list of the latest version of each solver
+    latest_solvers: Dict[str, Solver] = {}
+    for service in available_services:
+        if service["type"] == "computational":
```

Comment:
Pleas consider the following pattern to reduce nesting

```python
if service["type"] != "computational":
    continue
```

Reply:
I wrote
```python
    for service in available_services:
        if service["type"] == "computational":
               # do something ...
```
and you propose instead

```python
    for service in available_services:
        if service["type"] != "computational":
               continue
        # do something ...
```
You mean this? If so, can you elaborate on what are the extra benefits?

Reply:
it's about nesting.

Reply:
@pcrespov yes that is correct